lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/files/JobJarMaker.java

```diff
@@ -96,55 +102,88 @@ private static File toFile(final FileResource fileProto) {
   public File createJobSubmissionJAR(
       final JobSubmissionEvent jobSubmissionEvent,
       final Configuration driverConfiguration) throws IOException {
-    return createJAR(
-        driverConfiguration,
-        jobSubmissionEvent.getGlobalFileSet(),
-        jobSubmissionEvent.getLocalFileSet(),
-        this.fileNames.getDriverConfigurationName());
+
+    return newBuilder()
+        .withConfiguration(driverConfiguration)
+        .withGlobalFileSet(jobSubmissionEvent.getGlobalFileSet())
+        .withLocalFileSet(jobSubmissionEvent.getLocalFileSet())
+        .withConfigurationFileName(this.fileNames.getDriverConfigurationName())
+        .build();
   }
 
-  public File createJAR(
-      final Configuration configuration,
-      final Set<FileResource> globalFileSet,
-      final Set<FileResource> localFileSet,
-      final String configurationFileName) throws IOException {
+  private File makejobSubmissionFolder() throws IOException {
+    return Files.createTempDirectory(this.fileNames.getJobFolderPrefix()).toFile();
+  }
 
-    // Copy all files to a local job submission folder
-    final File jobSubmissionFolder = makejobSubmissionFolder();
-    LOG.log(Level.FINE, "Staging submission in {0}", jobSubmissionFolder);
+  /**
+   * Builder class for building JAR files.
+   */
+  public final class JarBuilder {
 
-    final File localFolder = new File(jobSubmissionFolder, this.fileNames.getLocalFolderName());
-    final File globalFolder = new File(jobSubmissionFolder, this.fileNames.getGlobalFolderName());
+    private Set<FileResource> localFiles = new HashSet<>();
+    private Set<FileResource> globalFiles = new HashSet<>();
+    private Configuration configuration = null;
+    private String configurationFilename = null;
 
-    copy(globalFileSet, globalFolder);
-    copy(localFileSet, localFolder);
+    private JarBuilder() {}
 
-    // Store the Configuration in the JAR file.
-    this.configurationSerializer.toFile(
-        configuration, new File(localFolder, configurationFileName));
+    public JarBuilder withLocalFileSet(final Set<FileResource> withLocalFiles) {
```

Comment:
Having below should be better since in line 123 and 124 you already initialized the Set.
public JarBuilder addLocalFileSet(final Set<FileResource> withLocalFiles)
public JarBuilder addLocalFile(FileResource localFile)

Reply:
The reason I named the methods "with*" is because I was following Azure Batch Java SDK builder naming conventions. Also, the verb "add" makes it seem that the method will add the given set to the existing set.